text/0026-semantic-versioning.md

```diff
@@ -0,0 +1,109 @@
+>   - MINOR version when you add functionality in a backwards compatible manner, and
+>   - PATCH version when you make backwards compatible bug fixes.
+
+The proposal is to conform to the above semantics as much as possible.
+
+### Changes
+
+In terms of Paketo Buildpacks/Builders, this translates to:
+
+* Bumping MAJOR is analogous to an "`app or related buildpacks may not work without change`" relation between releases.
+* Bumping MINOR is analogous to an "`updates or adds features`" relation between releases.
+* Bumping PATCH is analogous to a "`fixes issue`" relation between releases.
+
+The following table lists the usual circumstances/events where bumping a particular component is expected:
+
+#### Buildpack
+| `Major` | `Minor` | `Patch` |
+|-|-|-|
+| \<Any incompatible change\> | \<Feature additions but compatible with previous release\> | \<Fixes but no major feature changes\> |
+| Change in Buildpack ID | Addition/removal of dependency<br>(but default-version not changed to a version with breaking change) | Bug fixes |
+| Change in supported Stacks | Buildpack runs a new command the causes backward compatible<br>change in behavior (e.g. x-install buildpack adds running "x clean") | Performance improvement, internal tweaks |
+| Change in API (Provides/Requires) | Change of file type of build/detect executable<br>(e.g. text to binary) | CVE fixes |
+| Default version of dependency bumped to incompatible version<br>(e.g. php7 to php8) | New configuration option (e.g. new environment variable support) | Change of required build tool version to package buildpack<br>(e.g. go version) |
+| Removal of configuration options<br>(e.g. buildpack.yml) | Change in Buildpack API version |  |
+| Removal of publishing to a certain image registry<br>(e.g. gcr.io, dockerhub) | Publishing to a new image registry |  |
+|  | Change in <br>[buildpack-specific metadata](https://github.com/buildpacks/spec/blob/buildpack/v0.6/buildpack.md#build-plan-toml) |  |
+
+#### Buildpack (Language-family specific)
+| `Major` | `Minor` | `Patch` |
+|-|-|-|
+| Implementation buildpack additions/removals/major bumps<br>that cause incompatible change |  |  |
+
+
+#### Builder
+| `Major`                 | `Minor`                   | `Patch`                   |
+|-------------------------|---------------------------|---------------------------|
+| Lifecycle major bumps   | Minor bumps of components | Patch bumps of components |
+| Stack Image major bumps |                           |                           |
+| Buildpack major bumps   |                           |                           |
+
+### Intial Development
+
+A version of `0.0.x` is considered to be in initial development and may not be suitable for
+user consumption.
+
+## User Policy
+
+* **Patch** bump: User should update their buildpack/builder without hesitation.
+
+* **Minor** bump: User can update their buildpack/builder to avail new features. Nothing will break.
```

Comment:
For the "User Policy", I wonder if this statement is true "User can update their buildpack/builder to avail new features. Nothing will break."

Since one of the acceptable buildpack changes for a minor bump include:
> Addition/removal of dependency
(but default-version not changed to a version with breaking change)

I think this would easily break applications with specified dependency versions. For instance, Gemfiles specify the version of bundler to use (sometimes down to the patch) and if that dependency version is removed, the application will fail to build with that buildpacks latest minor update.

Reply:
Good point. The policy of "**Nothing** will break" on minor bumps might be a little too much of a claim. Also, major bump on every dependency update sounds like a lot of noise to the user. Maybe something along the lines of `User can update their buildpack/builder to avail new features and dependencies. Major lines of dependency versions won't be removed`? Removal of major line of versions can be added as an example to the major column in examples.

@kvedurmu Curious to hear if you have thoughts?

Reply:
+1 I think this seems a little strong. You should be able to upgrade without problems if you're using default versions or following good hygiene (i.e. using a wildcard like `2.*`).

Personally, I don't think we should encourage version pinning to minor and definitely not patch releases, so to adjust this policy to accommodate someone pinning bundler to version 1.2.3 seems like a bad idea to me. Buildpacks should promote good version hygiene, like using the latest versions and if you're following that process you should be rewarded with frustration-free upgrades. If you decided to pin your minor/patch versions, then you opt-in to more work to do additional testing and maintenance on buildpack updates. My $0.02.

Reply:
Although the `Gemfile.lock` does show a version number that is specific down to the patch, Bundler itself is loose with the specifics of what versions can be used. The `bundler` buildpack will install the latest version of Bundler matching the same major version as what is specified in the `Gemfile.lock`: https://github.com/paketo-buildpacks/bundler/blob/498cc180ad03991d79aa9fbf0ec2ec3391d144c6/gemfile_lock_parser.go#L18-L43. So, unless you are explicitly pinning a specific version number using either `buildpack.yml` or `BP_BUNDLER_VERSION`, you should not see a breaking change in this case.